scripts/cleanup_datasets/pgcleanup.py

```diff
@@ -844,17 +850,66 @@ class PurgeDeletedHDAs(PurgesHDAs, RemovesMetadataFiles, RequiresDiskUsageRecalc
     )
 
 
+class PurgeOldHDAs(PurgesHDAs, RemovesMetadataFiles, RequiresDiskUsageRecalculation, Action):
+    """
+    - Mark purged all HistoryDatasetAssociations that are older than the specified number of days.
+    - Mark deleted all MetadataFiles whose hda_id is purged in this step.
+    - Mark deleted all ImplicitlyConvertedDatasetAssociations whose hda_parent_id is purged in this
+      step.
+    - Mark purged all HistoryDatasetAssociations for which an ImplicitlyConvertedDatasetAssociation
+      with matching hda_id is deleted in this step.
+    """
+
+    force_retry_sql = " AND NOT history_dataset_association.purged"
+    _action_sql = """
+        WITH purged_hda_ids
+          AS (     UPDATE history_dataset_association
+                      SET purged = true, deleted = true{update_time_sql}
+                    FROM dataset
```

Comment:
Is this join implicit without something [like this](https://github.com/galaxyproject/galaxy/blob/e66075101eab2c8dbd3d334451c15cfeeb14f497/scripts/cleanup_datasets/pgcleanup.py#L896)? I've never understood exactly how implicit joins work.

Reply:
This was probably a bug you just caught - I've updated that line and the tests still pass so... that is a bit scary.